tests/src/runtimeApi/stream/StreamAddCallback.cpp

```diff
@@ -0,0 +1,117 @@
+#include <stdio.h>
+#include <hip/hip_runtime.h>
+#include <unistd.h>
+#include "test_common.h"
+
+/* HIT_START
+ * BUILD: %t %s ../../test_common.cpp
+ * TEST: %t
+ * HIT_END
+ */
+
+struct UserData
+{
+    size_t size;
+    int* ptr;
+};
+
+void myCallback(hipStream_t stream, hipError_t status, void* user_data)
```

Comment:
Is this actually automatically testing anything? I appreciate that this is likely visually showing that the two printfs in the callback should happen before anything after the streamsync, but the expected behavior is not described anywhere in the test code, and the test itself does not ensure that this happens. If the callback fails to actually block the streamsync, this test will pass and no one will notice until they go back and specifically look at the order of prints (assuming they know this is what they're looking for).

I think this test could be augmented by having the functions atomically increment some global variables, and ensuring that the resulting values do not indicate that the functions ran out of order.

For instance:
a global atomic_int callback_val.
At the top of test(), atomic_set(&callback_val, 0)

Before the first printf and after the second printf in the callback, fetch_add(&callback_val, 1)

After the hipDeviceSynchronize in test(), do an atomic_load(&callback) and verify that the value is 2. If it's <2, then that means that you have passed the streamsync without having completed the entirety of the callback.

This would allow you to return false from the test if the callback doesn't function correctly, not just if the copies fail for some reason.

Reply:
Agree